src/crowdin_client.cpp

```diff
@@ -135,150 +152,329 @@ CrowdinClient::~CrowdinClient() {}
+                i.pathName = str::to_wstring(pathStr) + i.pathName;
+        }
+
+        //TODO: get more until all branches gotten (if more than 500)    
+        return m_api->get(url + "/branches?limit=500");
+    }).then([this, url, prj](json r) {
+        // Handle branches
+        std::map<int, std::string> branches;
+
+        for(const auto& i : r["data"])
+        {
+            const json& d = i["data"];
+            branches[d["id"]] = d["name"].get<std::string>();
+        }
+
+        for(auto& i : prj->files)
+            if(i.branchId != NO_ID)
+                i.pathName = "/" + str::to_wstring(branches[i.branchId]) + i.pathName;
+
+        return *prj;
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::DownloadFile(const std::string& project_id,
-                                                   const std::wstring& file,
-                                                   const Language& lang,
+dispatch::future<void> CrowdinClient::DownloadFile(const long project_id,
+                                                   const long file_id,
+                                                   const std::wstring& file_name,
+                                                   const std::string& lang_tag,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+     wxLogTrace("poedit.crowdin", "Getting file URL: projects/%ld/translations/builds/files/%ld", project_id, file_id);
+    return RefreshToken().then([=] () {
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data({
+            { "targetLanguageId", lang_tag },
+            // for XLIFF files should be exported "as is" so set to `false`
+            { "exportAsXliff", !wxString(file_name).Lower().EndsWith(".xliff.xliff") }
+        }))
+        .then([this, output_file] (json r) {
+            std::string url = r["data"]["url"],
+                   proto(wxString(url).BeforeFirst(':').mb_str()),
+                   host(wxString(url).AfterFirst('/').AfterFirst('/').BeforeFirst('/').mb_str());
+
+            auto downloader = std::make_shared<crowdin_http_client>(*this, proto + "://" + host);
+            wxLogTrace("poedit.crowdin", "Gotten file URL: %s", r.dump().c_str());
+            return downloader->download(url, output_file)
+                .then([downloader]() {
+                    return dispatch::make_ready_future();
+                });
+        });
+    });
```

Comment:
Is this explicit `[downloader]` to make the object destroy itself appropriately (as you previously explained to me)? If yes, it warrants a comment so that it doesn't get removed as superfluous.

I don't understand why this extra nested future is being created and returned, though.

Reply:
Seems done all what you expected 199f93736a857f37a7a01438e9290a87558ca027, see below

> Is this explicit `[downloader]` to make the object destroy itself appropriately (as you previously explained to me)?

In short, yes.

>  If yes, it warrants a comment so that it doesn't get removed as superfluous.

Added

> I don't understand why this extra nested future is being created and returned, though.

Indeed builds and works without it :). Removed